File: apps/config-ui/src/stores/pointTable.js
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { invoke } from '@tauri-apps/api/core'
import { ElMessage } from 'element-plus'

export const usePointTableStore = defineStore('pointTable', () => {
  const tables = ref([])
  const currentTable = ref(null)
  const loading = ref(false)
  const error = ref(null)

  const tableCount = computed(() => tables.value.length)

  async function fetchTables() {
    loading.value = true
    error.value = null
    try {
      tables.value = await invoke('get_point_tables')
    } catch (e) {
      error.value = e.toString()
      ElMessage.error(`获取点表列表失败: ${e}`)
    } finally {
      loading.value = false
    }
  }

  async function fetchTable(id) {
    loading.value = true
    error.value = null
    try {
      currentTable.value = await invoke('get_point_table', { id })
    } catch (e) {
      error.value = e.toString()
      ElMessage.error(`获取点表失败: ${e}`)
    } finally {
      loading.value = false
    }
  }

  async function createTable(name, protocolType) {
    loading.value = true
    error.value = null
    try {
      const newTable = await invoke('create_point_table', { name, protocolType })
      await fetchTables()
      ElMessage.success('点表创建成功')
      return newTable
    } catch (e) {
      error.value = e.toString()
      ElMessage.error(`创建点表失败: ${e}`)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function deleteTable(id) {
    loading.value = true
    error.value = null
    try {
      await invoke('delete_point_table', { id })
      await fetchTables()
      ElMessage.success('点表删除成功')
    } catch (e) {
      error.value = e.toString()
      ElMessage.error(`删除点表失败: ${e}`)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function uploadCsv(tableId, csvType, content) {
    loading.value = true
    error.value = null
    try {
      const result = await invoke('upload_csv_file', { tableId, csvType, content })
      if (result.is_valid) {
        ElMessage.success(`CSV文件上传成功`)
        if (result.warnings.length > 0) {
          result.warnings.forEach(warning => {
            ElMessage.warning(`警告: ${warning.message}`)
          })
        }
      } else {
        result.errors.forEach(error => {
          ElMessage.error(`错误: ${error.message}`)
        })
      }
      await fetchTable(tableId)
      return result
    } catch (e) {
      error.value = e.toString()
      ElMessage.error(`上传CSV失败: ${e}`)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function exportCsv(tableId, csvType) {
    loading.value = true
    error.value = null
    try {
      const content = await invoke('export_csv_file', { tableId, csvType })
      return content
    } catch (e) {
      error.value = e.toString()
      ElMessage.error(`导出CSV失败: ${e}`)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function validateTable(tableId) {
    loading.value = true
    error.value = null
    try {
      const result = await invoke('validate_point_table', { tableId })
      if (result.is_valid) {
        ElMessage.success('点表验证通过')
      } else {
        ElMessage.error(`点表验证失败，发现 ${result.errors.length} 个错误`)
      }
      return result
    } catch (e) {
      error.value = e.toString()
      ElMessage.error(`验证点表失败: ${e}`)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function updatePoint(tableId, pointType, pointId, pointData) {
    loading.value = true
    error.value = null
    try {
      await invoke('update_point', { tableId, pointType, pointId, pointData })
      await fetchTable(tableId)
      ElMessage.success('点位更新成功')
    } catch (e) {
      error.value = e.toString()
      ElMessage.error(`更新点位失败: ${e}`)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function deletePoint(tableId, pointType, pointId) {
    loading.value = true
    error.value = null
    try {
      await invoke('delete_point', { tableId, pointType, pointId })
      await fetchTable(tableId)
      ElMessage.success('点位删除成功')
    } catch (e) {
      error.value = e.toString()
      ElMessage.error(`删除点位失败: ${e}`)
      throw e
    } finally {
      loading.value = false
    }
  }

  async function exportToComsrv(tableId) {
    loading.value = true
    error.value = null
    try {
      const path = await invoke('export_to_comsrv_format', { tableId })
      ElMessage.success(`导出成功: ${path}`)
      return path
    } catch (e) {
      error.value = e.toString()
      ElMessage.error(`导出失败: ${e}`)
      throw e
    } finally {
      loading.value = false
    }
  }

  return {
    tables,
    currentTable,
    loading,
    error,
    tableCount,
    fetchTables,
    fetchTable,
    createTable,
    deleteTable,
    uploadCsv,
    exportCsv,
    validateTable,
    updatePoint,
    deletePoint,
    exportToComsrv
  }
})

File: apps/config-ui/src-tauri/src/commands/point_table.rs
use crate::models::point_table::*;
use crate::services::point_table_service::PointTableService;
use std::sync::Arc;
use tauri::State;

#[tauri::command]
pub async fn get_point_tables(
    service: State<'_, Arc<PointTableService>>,
) -> std::result::Result<Vec<PointTableMetadata>, String> {
    service
        .get_all_tables()
        .await
        .map_err(|e| format!("Failed to get point tables: {}", e))
}

#[tauri::command]
pub async fn get_point_table(
    service: State<'_, Arc<PointTableService>>,
    id: String,
) -> std::result::Result<PointTable, String> {
    service
        .get_table(&id)
        .await
        .map_err(|e| format!("Failed to get point table: {}", e))
}

#[tauri::command]
pub async fn create_point_table(
    service: State<'_, Arc<PointTableService>>,
    name: String,
    protocol_type: String,
) -> std::result::Result<PointTableMetadata, String> {
    service
        .create_table(name, protocol_type)
        .await
        .map_err(|e| format!("Failed to create point table: {}", e))
}

#[tauri::command]
pub async fn delete_point_table(
    service: State<'_, Arc<PointTableService>>,
    id: String,
) -> std::result::Result<(), String> {
    service
        .delete_table(&id)
        .await
        .map_err(|e| format!("Failed to delete point table: {}", e))
}

#[tauri::command]
pub async fn upload_csv_file(
    service: State<'_, Arc<PointTableService>>,
    table_id: String,
    csv_type: CsvType,
    content: String,
) -> std::result::Result<ValidationResult, String> {
    service
        .upload_csv(&table_id, csv_type, content)
        .await
        .map_err(|e| format!("Failed to upload CSV: {}", e))
}

#[tauri::command]
pub async fn export_csv_file(
    service: State<'_, Arc<PointTableService>>,
    table_id: String,
    csv_type: CsvType,
) -> std::result::Result<String, String> {
    service
        .export_csv(&table_id, csv_type)
        .await
        .map_err(|e| format!("Failed to export CSV: {}", e))
}

#[tauri::command]
pub async fn validate_point_table(
    service: State<'_, Arc<PointTableService>>,
    table_id: String,
) -> std::result::Result<ValidationResult, String> {
    service
        .validate_table(&table_id)
        .await
        .map_err(|e| format!("Failed to validate point table: {}", e))
}

#[tauri::command]
pub async fn update_point(
    service: State<'_, Arc<PointTableService>>,
    table_id: String,
    point_type: String,
    point_id: u32,
    point_data: serde_json::Value,
) -> std::result::Result<(), String> {
    service
        .update_point(&table_id, &point_type, point_id, point_data)
        .await
        .map_err(|e| format!("Failed to update point: {}", e))
}

#[tauri::command]
pub async fn delete_point(
    service: State<'_, Arc<PointTableService>>,
    table_id: String,
    point_type: String,
    point_id: u32,
) -> std::result::Result<(), String> {
    service
        .delete_point(&table_id, &point_type, point_id)
        .await
        .map_err(|e| format!("Failed to delete point: {}", e))
}

#[tauri::command]
pub async fn export_to_comsrv_format(
    service: State<'_, Arc<PointTableService>>,
    table_id: String,
) -> std::result::Result<String, String> {
    service
        .export_to_comsrv_format(&table_id)
        .await
        .map_err(|e| format!("Failed to export to comsrv format: {}", e))
}

#[tauri::command]
pub async fn get_protocol_csv_template(
    service: State<'_, Arc<PointTableService>>,
    protocol_type: String,
    csv_type: CsvType,
) -> std::result::Result<String, String> {
    service
        .get_protocol_csv_template(&protocol_type, csv_type)
        .await
        .map_err(|e| format!("Failed to get CSV template: {}", e))
}